File: client/src/reducers/material.js
import * as actions from '../constants/materialActionTypes';

const materialReducer = (state = {}, action) => {
  const { type, payload, meta } = action;

  switch (type) {
    case `${actions.FETCH_MATERIALS}_FULFILLED`:
      return {
        ...state,
        ...payload.reduce((acc, curr) => {
          acc[curr.id] = {
            id: curr.id,
            name: curr.name,
            number: curr.number,
            count: curr.count,
            expirationDate: curr.expiration_date,
            price: curr.price,
            units: curr.units,
            dateCreated: curr.date_created,
            dateModified: curr.date_modified,
          };
          return acc;
        }, {}),
      };
    case `${actions.FETCH_MATERIAL}_FULFILLED`:
      return {
        ...state,
        [payload.id]: {
          ...payload,
          id: payload.id,
          name: payload.name,
          number: payload.number,
          count: payload.count,
          expirationDate: payload.expiration_date,
          price: payload.price,
          units: payload.units,
          dateCreated: payload.date_created,
          dateModified: payload.date_modified,
        },
      };

    case `${actions.CREATE_MATERIAL}_FULFILLED`: {
      const { material } = meta;

      return {
        ...state,
        [payload.id]: {
          ...material,
          id: payload.id,
          dateCreated: payload.date_created,
          dateModified: payload.date_modified,
        },
      };
    }

    case `${actions.UPDATE_MATERIAL}_FULFILLED`: {
      const { material } = meta;

      return {
        ...state,
        [material.id]: {
          ...state[material.id],
          ...material,
          dateModified: payload.date_modified,
        },
      };
    }

    case `${actions.DELETE_MATERIAL}_FULFILLED`: {
      const { id } = meta;
      const deleteState = Object.assign(state);
      delete deleteState[id];
      return deleteState;
    }

    default:
      return state;
  }
};

export default materialReducer;

File: server/chalicelib/material.py
# pylint: disable=relative-beyond-top-level
from graphene import Field, ID, relay, Float
from . import base, types

"""
+-----------------------------+-------------------------------------+------+-----+-------------------+----------------+
| Field                       | Type                                | Null | Key | Default           | Extra          |
+-----------------------------+-------------------------------------+------+-----+-------------------+----------------+
| id                          | int(11)                             | NO   | PRI | NULL              | auto_increment |
| number                      | varchar(255)                        | NO   |     | NULL              |                |
| notes                       | text                                | NO   |     | NULL              |                |
| date_created                | datetime                            | NO   |     | CURRENT_TIMESTAMP |                |
| date_modified               | datetime                            | NO   |     | CURRENT_TIMESTAMP |                |
| expiration_date             | date                                | YES  |     | NULL              |                |
| name                        | varchar(255)                        | NO   |     | NULL              |                |
| price                       | int(10) unsigned                    | NO   |     | NULL              |                |
| units                       | enum('UNIT','KG','LB','G','L','ML') | NO   |     | UNIT              |                |
| count                       | float unsigned                      | NO   |     | NULL              |                |
| purchase_order_url          | varchar(2083)                       | YES  |     | NULL              |                |
| purchase_order_number       | varchar(255)                        | YES  |     | NULL              |                |
| certificate_of_analysis_url | varchar(2083)                       | YES  |     | NULL              |                |
+-----------------------------+-------------------------------------+------+-----+-------------------+----------------+
"""


class MaterialBase(
    types.Numberable,
    types.Notable,
    types.Expirable,
    types.Namable,
    types.Pricable,
    types.Measurable,
    types.ContinuousCountable,
    types.HasVendor,

):
    pass


class TableName:
    __tablename__ = 'material'


class MaterialInput(base.Input, MaterialBase, TableName):

    def to_output(self, id, date_created, date_modified):
        return Material(
            id=id,
            date_created=date_created,
            date_modified=date_modified,
            **self.__dict__,
        )


class Material(base.Object, MaterialBase, types.Node, TableName):
    __tablename__ = 'material'

    class Meta:
        interfaces = (relay.Node,)


class MaterialConnection(base.ObjectConnection, TableName):

    class Meta:
        node = Material

    class Edge:
        count = types.Integer()

        @staticmethod
        def resolve_count(parent, info):
            print(parent)
            return parent.node['count'] if parent else None


class CreateMaterial(base.Create, TableName):
    __tablename__ = 'material'

    class Arguments:
        material = MaterialInput(required=True)

    material = Field(Material)

    @staticmethod
    def mutate(parent, info, material):
        return {'material': CreateMaterial.commit(material)}


class UpdateMaterial(base.Update, TableName):
    __tablename__ = 'material'

    class Arguments:
        id = ID(required=True)
        material = MaterialInput(required=True)

    material = Field(Material, required=True)

    @staticmethod
    def mutate(parent, info, id, material):
        return {'material': UpdateMaterial.commit(id, material)}


class DeleteMaterial(base.Delete, TableName):
    __tablename__ = 'material'

    class Arguments:
        id = ID(required=True)

    id = ID(required=True)

    @staticmethod
    def mutate(parent, info, id):
        DeleteMaterial.commit(id, usees=[], users=['product', 'case'])
        return {'id': id}
